### PR TITLE
[Stdlib] Use associated type where clauses to address various ABI FIXMEs.

### PR DESCRIPTION
**Pull request by DougGregor**
Start using associated type where clauses in the standard library to model same-type constraints that were inexpressible before [SE-0142](https://github.com/apple/swift-evolution/blob/master/proposals/0142-associated-types-constraints.md). Clean up the warnings that result from these constraints being in the protocols where they belong.

Resolves [SR-2121](https://bugs.swift.org/browse/SR-2121) and standard library ABI FIXMEs 68, 89, 90, 91, 92, 94, 96, and most of 99. Some of these are merely warnings in the unit testing library, but a few (68, 92, 99) are part of the core standard library protocols.

### CODE REVIEW

**Comment by DougGregor** [comment]
@dabrahams , @huonw , and @gribozavr might be interested in this. It's WIP right now because it regresses a compiler crasher

**Comment by slavapestov** [comment]
It's unfortunate that associatedtype _Element pops up in completion results now - but I see it has a FIXME saying it's a hack that should be removed. Any chance the problem is already fixed, or we could fix it pretty easily?

**Comment by DougGregor** [comment]
The problem is "the library needs recursive constraints"

**Comment by dabrahams** [comment]
@DougGregor Are you sure?  IIRC, `_Element' should be `Element' and should be constrained to be the same as `Iterator.Element` (and all the `Iterator.Element`s in constraints should be changed to `Element`).

**Comment by DougGregor** [comment]
Oh! @dabrahams has the right answer... that's nice big cleanup to have 

    associatedtype Element

within `Sequence` and have it be equivalent to (and, for expressiveness purposes, preferred over) `Iterator.Element`.

**Comment by DougGregor** [comment]
Just gotta stop that one crasher from looping ;)

**Comment by dabrahams** [comment]
Very W00t and Wunderful to see all that red in the diff!  Do let me know the moment we can start using it in the unicode-rethink branch!

**Comment by huonw** [comment]
I think the crasher Doug is struggling with is fairly invalid code that, even without this PR, currently still crashes, so, you should be able to use (/find bugs in) these where clauses now.

**Comment by DougGregor** [comment]
Yeah, my struggles are entirely with recursively-constructed types now, not with the rest of the feature. *That said*, minor perturbations of the Sequence/Collection protocols trip over the recursively-constructed types, so when you see the compiler spinning in `compareDependentTypes`.... you hit one.

**Comment by DougGregor** [comment]
The infinite recursion is here:

  https://github.com/apple/swift/blob/master/lib/AST/GenericSignatureBuilder.cpp#L2796

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: stdlib/public/core/Collection.swift [hunk 1]
```diff
@@ -646,6 +646,7 @@ public protocol Collection : _Indexable, Sequence {
   /// supplying `IndexingIterator` as its associated `Iterator`
   /// type.
   associatedtype Iterator = IndexingIterator<Self>
+    where Self.Iterator.Element == _Element
```
Comment (slavapestov) [inline stdlib/public/core/Collection.swift]: Could we rename _Element to Element, and write Element in place of Iterator.Element everywhere? :-)

**Comment by DougGregor** [comment]
@dabrahams suggested that above... I'm going to take a shot at it

**Pull request merged**